kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h

```diff
@@ -0,0 +1,714 @@
+                rExtensions.GetZeroDerivativesVariables(rVec, rCurrentProcessInfo);
+            },
+            r_current_process_info);
+
+#pragma omp parallel for
+        for (int i = 0; i < number_of_elements; ++i)
+        {
+            Element& r_element = *(rModelPart.ElementsBegin() + i);
+            Geometry<Node<3>>& r_geometry = r_element.GetGeometry();
+            SchemeExtension& r_extensions = *r_element.GetValue(SCHEME_EXTENSION);
+
+            const int k = OpenMPUtils::ThisThread();
+
+            for (unsigned int i_node = 0; i_node < r_geometry.PointsNumber(); ++i_node)
+            {
+                r_extensions.GetZeroDerivativesVector(
+                    i_node, mIndirectCurrentDisplacementVector[k], 0, r_current_process_info);
+                r_extensions.GetZeroDerivativesVector(
+                    i_node, mIndirectOldDisplacementVector[k], 1, r_current_process_info);
+                r_extensions.GetFirstDerivativesVector(
+                    i_node, mIndirectOldVelocityVector[k], 1, r_current_process_info);
+                r_extensions.GetSecondDerivativesVector(
+                    i_node, mIndirectCurrentAccelerationVector[k], 0, r_current_process_info);
+                r_extensions.GetSecondDerivativesVector(
+                    i_node, mIndirectOldAccelerationVector[k], 1, r_current_process_info);
+
+                r_geometry[i_node].SetLock();
+                for (unsigned int i_var = 0;
+                     i_var < mIndirectCurrentDisplacementVector[k].size(); ++i_var)
+                    mIndirectCurrentDisplacementVector[k][i_var] =
+                        mIndirectOldDisplacementVector[k][i_var] +
+                        mBossak.C6 * mIndirectOldVelocityVector[k][i_var] +
+                        mBossak.C4 * mIndirectOldAccelerationVector[k][i_var] +
+                        mBossak.C5 * mIndirectCurrentAccelerationVector[k][i_var];
+                r_geometry[i_node].UnSetLock();
+            }
+        }
+        SynchronizeVariables(displacement_vars, rModelPart.GetCommunicator());
+    }
+
+    struct Hash
+    {
+        std::size_t operator()(const VariableData* const& p) const
+        {
+            return p->Key();
+        }
+    };
+
+    struct Pred
+    {
+        bool operator()(const VariableData* const l, const VariableData* const r) const
+        {
+            return *l == *r;
+        }
+    };
+
+    // Gathers variables needed for assembly.
+    static std::vector<const VariableData*> GatherVariables(
+        const ModelPart::ElementsContainerType& rElements,
+        std::function<void(const SchemeExtension&, std::vector<const VariableData*>&, ProcessInfo&)> GetLocalVars,
+        ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+        const int num_threads = OpenMPUtils::GetNumThreads();
+        std::vector<const VariableData*> local_vars;
+        std::vector<std::unordered_set<const VariableData*, Hash, Pred>> thread_vars(num_threads);
```

Comment:
ouch, here i am really lost. what do you want to accomplish??

aside ot this it is better if your thread variables locally within a parallel region and then you insert them within a critical sectin

~~~cpp
#pragma omp parallel
{
     std::unordered_set<const VariableData*, Hash, Pred> local_var
     ...fill local var...
     #pragma omp critical
     all_vars.insert(local_var.begin(), local_var].end());
}
~~~